Update howto compare strings

BillWagner opened the pull request:
Update the "how to: compare strings" topic.

1. Update the samples to use more modern techniques.
1. break out the samples into smaller pieces.
1. Update the explanations for more modern library elements.
1. Re-order for easier explanation.

[Internal review link](https://review.docs.microsoft.com/en-us/dotnet/csharp/how-to/compare-strings?branch=pr-en-us-4596)

BillWagner committed:
initial move and redirect

move "How to compare strings" from the programming guide  to the how to section. Move links, TOC, and redirection.

BillWagner committed:
Samples update part 1.

Pull apart larger samples into smaller nuggets.

rpetrusha commented:
The build has failed, @BillWagner.

BillWagner commented:
Thanks @rpetrusha 

And, this isn't ready for a review yet. I was doing a test build on the samples.

BillWagner committed:
fix build errors

BillWagner committed:
more build errors

BillWagner committed:
fix one more build error

BillWagner committed:
update link.

BillWagner committed:
update contents

Create zip file for sample.

BillWagner committed:
proofread and update

Found a couple mistakes with sample snippets.
Fixed grammar and tone.

BillWagner commented:
@rpetrusha This is now ready for review.

rpetrusha commented on docs/csharp/how-to/compare-strings.md:
```diff
@@ -0,0 +1,86 @@
+
+[!code-csharp-interactive[Sorting an array of strings](../../../samples/snippets/csharp/how-to/strings/CompareStrings.cs#5)]
+
+Once the array is sorted, you can search for entries using a binary search. A binary search starts in the middle of the collection to determine which half of the collection would contain the sought string. Each subsequent comparison subdivides the remaining part of the collection in half. The following sample demonstrates the importance of using the same comparison rules for both sorting and sear ...
```
would contain --> contains
It's not clear without careful inspection (especially since the output display the search string, not the string found) how the example demonstrates the importance of using the same comparison rules -- either mention explicitly that what's found is a partly uppercase string that's not an exact match for an array element, or change the value displayed, as I mentioned below in the code.

rpetrusha commented on samples/snippets/csharp/how-to/strings/CompareStrings.cs:
```diff
@@ -0,0 +1,255 @@
+            int result = Array.BinarySearch(lines, searchString, StringComparer.OrdinalIgnoreCase);
+            ShowWhere<string>(lines, result);
+
+            Console.WriteLine($"{(result > 0 ? "Found" : "Did not find")} {searchString}");
```
See my comment from the text.
Rather than showing the search string if it's found, it would be clearer to display the found string.

rpetrusha commented on docs/csharp/how-to/compare-strings.md:
```diff
@@ -0,0 +1,86 @@
+
+## Case sensitive and culture sensitive sorting and searching
+
+The following code uses the <xref:System.Collections.Generic.List%601?displayProperty=nameWithType> collection class to store strings. The strings are sorted using the <xref:System.Collections.Generic.List%601.Sort%2A?displayProperty=nameWithType> method. This method needs a delegate that compares and orders two strings. The <xref:System.String.CompareTo%2A?displayProperty=nameWithType> method pr ...
```
Here again, observing the order is complicated by output being in a scrollable window. Perhaps include output from the example.

rpetrusha commented on docs/csharp/how-to/compare-strings.md:
```diff
@@ -0,0 +1,86 @@
+
+[!code-csharp-interactive[Sorting a list of strings](../../../samples/snippets/csharp/how-to/strings/CompareStrings.cs#7)]
+
+Once sorted, the list of strings can be searched using a binary search. The following sample shows how to search the sorted listed using the same comparison function. The local function `ShowWhere` shows where the sought text is or would be:
```
Binary search is not appropriate here. If you're doing a binary search on culture-sensitive strings, searches are likely to fail for strings that contain characters outside the ASCII range. The search method should correspond to the sorting method.

BillWagner replied:
@rpetrusha This is the one comment that confuses me. What search method should be used when the collection is sorted in a culture-sensitive manner? Is it necessary to do a linear search?

rpetrusha commented on docs/csharp/how-to/compare-strings.md:
```diff
@@ -0,0 +1,86 @@
+
+The following example shows how to compare strings using the <xref:System.String?displayProperty=nameWithType> methods that take a <xref:System.StringComparison> enumeration. The <xref:System.String.CompareTo%2A?displayProperty=nameWithType> instance methods are not used here, because overloads take a <xref:System.StringComparison>. Restrict your use of the <xref:System.String.CompareTo%2A?displa ...
+
+[!code-csharp-interactive[Comparing strings across cultures](../../../samples/snippets/csharp/how-to/strings/CompareStrings.cs#9)]
```
There's a problem with this code in the REPL. Here's the output:
     First sentence is <Sie tanzen auf der Straße.>
     Second sentence is <Sie tanzen auf der Strasse.>
     Comparing in Invariant Language (Invariant Country) returns 1.
     Comparing in de-DE returns 1.
     The two strings are not equal.
The issues:
- For some reason, en-US was not set as the current culture, even though it should have been.
- Even the de-DE comparison fails. The basic problem may be that the REPL is running on a non-Windows machine. Windows supplies the culture-sensitive character tables for .NET Core (I think). I have no idea of where culture-sensitive character comes from on non-Windows systems, though. @tarekgh, might that be the source of the problem?

tarekgh replied:
The string comparisons and searching operations depend on the OS, the framework doesn't carry the collation tables and it just call the underlying OS for such operations. so the behavior is fully depended on the OS.

rpetrusha reviewed (changes requested):
This looks good, @BIllWagner. I've left a number of comments, but most concerning is that the REPL doesn't behave as expected for the culture-sensitive de-DE comparison. @tarekgh, could you take a look at the example and the output, please?

rpetrusha commented on samples/snippets/csharp/how-to/strings/CompareStrings.cs:
```diff
@@ -0,0 +1,255 @@
+            // For culture-sensitive comparisons, use the String.Compare 
+            // overload that takes a StringComparison value.
+            int i = String.Compare(first, second, StringComparison.CurrentCulture);
+            Console.WriteLine($"Comparing in {originalCulture.EnglishName} returns {i}.");
```
I missed this in my original review. The comparison is with the current culture, which here is en-US. But the Console.WriteLine statement is reporting that it's the result for the original culture, which is CultureInfo.InvariantCulture. So that needs too  be changed. But even so, the comparison results are unexpected.

pkulikov commented on docs/csharp/how-to/compare-strings.md:
```diff
@@ -0,0 +1,86 @@
+
+When you compare strings, you define an order among them. Comparisons are used to sort a sequence of strings. Once the sequence is in a known order, it is easier to search, both for software and for humans. Other comparisons may check if strings are the same. These sameness checks are similar to equality, but some differences, such as case differences may be ignored.
+
+The methods that compare strings indicate that one string is greater than or less than the other, or that the two strings are equal. The comparison result depends on rules that specify performing an *ordinal comparison* or a *culture-sensitive comparison* and if case should be considered. You pick the rules that match your scenario.
```
What about variant: *one string precedes or follows the other* instead of *is greater than or less than* (I guess, swapping relations is fine). I'm inspired by phrasing in [`IComparable` reference](https://docs.microsoft.com/en-us/dotnet/api/system.icomparable-1.compareto?view=netframework-4.7.1)

pkulikov commented on docs/csharp/how-to/compare-strings.md:
```diff
@@ -0,0 +1,86 @@
+
+The methods that compare strings indicate that one string is greater than or less than the other, or that the two strings are equal. The comparison result depends on rules that specify performing an *ordinal comparison* or a *culture-sensitive comparison* and if case should be considered. You pick the rules that match your scenario.
+
+Use basic ordinal comparisons when you have to compare two strings without regard to linguistic conventions. A basic ordinal comparison (<xref:System.StringComparison.Ordinal?displayProperty=nameWithType>) is case-sensitive, which means that the two strings must match character for character: "and" does not equal "And" or "AND". A frequently used variation is <xref:System.StringComparison.Ordinal ...
```
One can read the term *compare* in two ways (@BillWagner  you already mentioned that in the first paragraph):
- figure out ordering of two strings
- figure out if two strings are equal or not

If a paragraph talks mostly about one meaning, I suggest to specify that.
Here, it looks, we talk only about *equal/not equal* meaning. As for sorting, I'm surprised to find that ordering changes:
```csharp
Console.WriteLine(string.Compare("and", "And", StringComparison.Ordinal));           // gives 32
Console.WriteLine(string.Compare("and", "And", StringComparison.InvariantCulture));  // gives -1
```

pkulikov commented on docs/csharp/how-to/compare-strings.md:
```diff
@@ -0,0 +1,86 @@
+
+Use basic ordinal comparisons when you have to compare two strings without regard to linguistic conventions. A basic ordinal comparison (<xref:System.StringComparison.Ordinal?displayProperty=nameWithType>) is case-sensitive, which means that the two strings must match character for character: "and" does not equal "And" or "AND". A frequently used variation is <xref:System.StringComparison.Ordinal ...
+
+Culture-sensitive comparisons are typically used to compare and sort strings that are input by users. The characters and sorting conventions of these strings might vary depending on the locale of the user's computer. Even strings that contain identical characters might sort differently depending on the culture of the current thread.
```
*The characters...* should be *The comparison...* at the start of the second sentence?

pkulikov commented on docs/csharp/how-to/compare-strings.md:
```diff
@@ -0,0 +1,86 @@
+
+## Comparing across cultures
+
+The following example shows several aspects of comparing strings across cultures. A string is initialized to the phrase "They dance in the street." in German. Linguistically (in Windows), "ss" is equal to the German essetz: 'ß' character in both "en-US" and "de-DE" cultures. This sample stores <xref:System.Globalization.CultureInfo> for the current culture. The original culture can be set and ret ...
```
Should *essetz* be *Eszett* (based on [wiki page](https://en.wikipedia.org/wiki/%C3%9F))?

pkulikov commented on docs/csharp/programming-guide/strings/index.md:
```diff
@@ -119,8 +119,7 @@ string s = String.Empty;
 |[How to: Modify String Contents](../../how-to/modify-string-contents.md)|Illustrates techniques to transform strings and modify the contents of strings.|  
-|[How to: Concatenate Multiple Strings](../../../csharp/programming-guide/strings/how-to-concatenate-multiple-strings.md)|Illustrates techniques to combine multiple strings into one string at compile time and run time.|  
-|[How to: Compare Strings](../../../csharp/programming-guide/strings/how-to-compare-strings.md)|Shows how to perform ordinal comparisons of strings.|  
+|[How to: Compare Strings](../../how-to/compare-strings.md)|Shows how to perform ordinal comparisons of strings.|  
```
Not only ordinal, cultures are also covered in the topic.

pkulikov commented on samples/snippets/csharp/how-to/strings/CompareStrings.cs:
```diff
@@ -0,0 +1,255 @@
+
+            thread.CurrentCulture = originalCulture;
+            bool b = String.Equals(first, second, StringComparison.CurrentCulture);
+            Console.WriteLine($"The two strings {(b == true ? "are" : "are not")} equal.");
```
Let's rename `b` to `areStringsEqual` or better name in order to replace `b == true` with just `areStringsEqual`

tarekgh commented on samples/snippets/csharp/how-to/strings/CompareStrings.cs:
```diff
@@ -0,0 +1,255 @@
+
+                    Console.Write("Not found. Sorts between: ");
+
+                    if (index == 0)
```
>if (index == 0) [](start = 20, length = 15)

Is this possible to have index == 0?

pkulikov replied:
@tarekgh `Array.BinarySearch` might return `-1`, then `index == 0` is `true`: note two lines above bitwise complement:
```csharp
index = ~index
```

tarekgh replied:
ok, thanks.

tarekgh commented on samples/snippets/csharp/how-to/strings/CompareStrings.cs:
```diff
@@ -0,0 +1,255 @@
+
+                    Console.Write("Not found. Sorts between: ");
+
+                    if (index == 0)
```
ditto

tarekgh commented on samples/snippets/csharp/how-to/strings/CompareStrings.cs:
```diff
@@ -0,0 +1,255 @@
+            Console.WriteLine($"Second sentence is <{second}>");
+
+            System.Threading.Thread thread = System.Threading.Thread.CurrentThread;
+            System.Globalization.CultureInfo originalCulture = thread.CurrentCulture;
```
I am not sure if this code depends on specific version of the framework. if this for net core or after 4.6 in the full framework, I would write this simply as

 System.Globalization.CultureInfo originalCulture = CultureInfo.CurrentCulture;

and setting current culture would be 

CultureInfo.CurrentCulture = ew System.Globalization.CultureInfo("en-US");

pkulikov commented on samples/snippets/csharp/how-to/strings/CompareStrings.cs:
```diff
@@ -0,0 +1,255 @@
+                        Console.Write($"{collection[index - 1]} and ");
+
+                    if (index == collection.Count)
+                        Console.WriteLine("end of array.");
```
It's not array anymore, it's a list.
The same true for *beginning of array* several lines above.

tarekgh replied:
I am not sure which comment you are replying to.

tarekgh commented on docs/csharp/how-to/compare-strings.md:
```diff
@@ -0,0 +1,86 @@
+
+## Ordinal comparison examples
+
+The following example shows how to correctly compare strings whose values will not change based on the locale of the user's computer. It declares and initializes two string variables. These strings are declared in code, so they will not be localized. The two strings are compared for equality using an ordinal comparison. An ordinal comparison is the fastest method to compare strings. Try it yourse ...
```
If you want to have first and second be equals on Linux runs, you will need to change the code to use something like 

```C#
System.Globalization.CultureInfo de = new System.Globalization.CultureInfo("de-DE");
Console.WriteLine(ci.CompareInfo.Compare(first, second, CompareOptions.IgnoreNonSpace));
```

BillWagner replied:
@rpetrusha said: I'm not sure what "they will not be localized" means.

I wasn't either. It was in the original code sample. I removed it. It didn't add anything to the discussion.

tarekgh replied:
@BillWagner for fixing the code. I don't have any more comments regarding the samples code. LGTM

BillWagner commented on docs/csharp/how-to/compare-strings.md:
```diff
@@ -0,0 +1,86 @@
+author: "BillWagner"
+ms.author: "wiwagn"
+---
+# How to compare strings in C# #
```
I had that. I moved it earlier in the text.

BillWagner committed:
respond to feedback

part 1.

Ron's feedback is making think more about how this is presented. I may be re-organizing the topic more.

BillWagner committed:
code updates

BillWagner committed:
interim checkin.

Having an offline discussion about the direction.

BillWagner committed:
major rework of the story and the samples

I think this is clearer, but I want to review carefully and check previous comments before requesting a new review.

BillWagner committed:
check comments.

Checking in to switch machines. Need to finish with the output on the windows machines.

BillWagner committed:
Platform updates for culture comparisons

BillWagner commented:
I've addressed all the review comments except for [this one](https://github.com/dotnet/docs/pull/4596#discussion_r174867766)  @rpetrusha Do you have guidance there?

rpetrusha commented on docs/csharp/how-to/compare-strings.md:
```diff
@@ -0,0 +1,163 @@
+
+## Linguistic sorting and searching strings in arrays
+
+The following examples show how to sort and search for strings in an array using a linguistic comparison dependent on the curren cultuer. You use the static <xref:System.Array> methods that take a <xref:System.StringComparer?displayProperty=nameWithType> parameter.
```
typos in "current culture" in the first sentence

rpetrusha commented:
In terms of my binary search is not appropriate, @BillWagner, I'd looked at the code too hastily, unfortunately, and I'd missed that the search was culture-sensitive rather than ordinal. I apologize. So the text and the example are fine. Except that the paragraph contains a typo, which I've indicated in a comment. Other than that, this looks really good and is ready to merge.

BillWagner committed:
fix typo.
